### DISCUSSION TITLE
[Flows] Send Mail with header

### DISCUSSION THREAD

**Discussion by LasseRosenow** [directus/directus#16726] [Feature Requests]
I am implementing a contact form with flows:

I have a collection with following fields:
- first_name
- last_name
- e_mail
- message

Whenever a new item gets created in this collection I want to send an e-mail. This is working so far.

But it would be very good to be able to set the "Reply-To" Header.

But I don't see an option in the flows UI.

What I need is to set this dynamically based on a payload field, so that the reply-to header directs back to the one who created a contact form entry.

**Comment by azrikahar** [comment]
Perhaps this can be worked on in tandem with #15913 to add them both (or more) as advanced options as mentioned in https://github.com/directus/directus/discussions/15913#discussioncomment-3854220. Seems to be the `replyTo` option in Nodemailer based on what I can see from their docs: https://nodemailer.com/message/#more-advanced-fields

**Comment by raphourbe** [comment]
Hi there,

I wanted to understand how to create custom operations for Flow, so I designed one for you that appears to work 😄 

You need to create an extension for operation. See: https://docs.directus.io/extensions/creating-extensions.html to create the template. Please pick Javascript instead of Typescript for the following code. Or adapt it to your Typescript needs.

Then, in app.js

```jsx
export default {
  id: "custom-mail-operation",
  name: "Custom Mail Operation",
  icon: "mail",
  description: "Enhanced the Send Mail Operation",
  overview: ({ body, to, cc, replyTo, bcc, type, subject, template }) => [
    {
      label: "Subject",
      text: subject,
    },
    {
      label: "To",
      text: Array.isArray(to) ? to.join(", ") : to,
    },
    {
      label: "Cc",
      text: Array.isArray(cc) ? cc.join(", ") : cc,
    },
    {
      label: "Bcc",
      text: Array.isArray(bcc) ? bcc.join(", ") : bcc,
    },
    {
      label: "Reply to",
      text: replyTo,
    },
    {
      label: "Type",
      text: type || "markdown",
    },
    {
      label: "Body",
      text: body,
    },
    {
      label: "Template",
      text: template,
    },
  ],
  options: (panel) => {
    return [
      {
        field: "to",
        name: "To",
        type: "csv",
        meta: {
          width: "full",
          interface: "tags",
          options: {
            iconRight: "alternate_email",
          },
        },
      },
      {
        field: "cc",
        name: "Cc",
        type: "csv",
        meta: {
          width: "full",
          interface: "tags",
          options: {
            iconRight: "alternate_email",
          },
        },
      },
      {
        field: "bcc",
        name: "Bcc",
        type: "csv",
        meta: {
          width: "full",
          interface: "tags",
          options: {
            iconRight: "alternate_email",
          },
        },
      },
      {
        field: "replyTo",
        name: "Reply To",
        type: "string",
        meta: {
          width: "full",
          interface: "input",
          options: {
            iconRight: "alternate_email",
          },
        },
      },
      {
        field: "subject",
        name: "Subject",
        type: "string",
        meta: {
          width: "full",
          interface: "input",
          options: {
            iconRight: "title",
          },
        },
      },
      {
        field: "template",
        name: "Template",
        type: "string",
        meta: {
          width: "full",
          interface: "input",
          options: {
            iconRight: "title",
          },
        },
      },
      {
        field: "type",
        name: "Type",
        type: "string",
        schema: {
          default_value: "markdown",
        },
        meta: {
          interface: "select-dropdown",
          width: "half",
          options: {
            choices: [
              {
                text: "markdown",
                value: "markdown",
              },
              {
                text: "wysiwyg",
                value: "wysiwyg",
              },
            ],
          },
        },
      },
      {
        field: "body",
        name: "Body",
        type: "text",
        meta: {
          width: "full",
          interface:
            panel.type === "wysiwyg"
              ? "input-rich-text-html"
              : "input-rich-text-md",
        },
      },
    ];
  },
};
```

And in api.js
```jsx
import { marked } from "marked";
import sanitizeHTML from "sanitize-html";

/**
 * Render and sanitize a markdown string
 */
export function md(str) {
  return sanitizeHTML(marked(str));
}

export default {
  id: "custom-mail-operation",
  handler: async (
    { body, to, cc, replyTo, bcc, type, subject, template },
    { accountability, database, getSchema, services }
  ) => {
    const { MailService } = services;
    const mailService = new MailService({
      schema: await getSchema({ database }),
      accountability,
      knex: database,
    });

    // If you don't want to specify a template use this.
    if (template == "" || template == null)
      await mailService.send({
        html: type === "wysiwyg" ? body : md(body),
        to: to,
        cc: cc,
        bcc: bcc,
        replyTo: replyTo,
        subject: subject,
      });
    // If you want a template => you need to have {{ html }}
    // in your template
    else
      await mailService.send({
        to: to,
        cc: cc,
        bcc: bcc,
        replyTo: replyTo,
        subject: subject,
        template: {
          name: template !== null || template !== undefined ? template : "base",
          data: { html: type === "wysiwyg" ? body : md(body) },
        },
      });
  },
};
```

You then just need to run `npm run build` in the extension folder, and place the content of the /dist folder into your directus folder => /extension/operations/custom-enhanced-mail.

I added a template selection too. So, if you do not have a custom template for sending mails, let it empty. If you do, type the name of your template file (if you have `super-template.liquid` for example as a template file, type `super-tempate` in the text area). Be aware that the html will be put into the `{{html}}` of your template. So you need one.

**Comment by shipsides-sparknel** [comment]
hey folks.

 +1ing an idea not specifically for custom headers (though would absolutely be useful!), but adding bcc and replyTo support since it seems to be part of nodemailer out of the box.

it would be very useful for cloud users to at a minimum specify a replyTo since we cannot change the SMTP and sender.

(changing the sender email would understandably be a bridge too far and affect the send rate without SPF and DKIM.)

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
